packages/perspective-viewer-d3fc/src/js/tooltip/tooltip.js

```diff
@@ -0,0 +1,86 @@
+import {select} from "d3";
+import {getChartElement} from "../plugin/root";
+import tooltipTemplate from "../../html/tooltip.html";
+
+export function tooltip(selection, settings) {
+    const container = select(getChartElement(selection.node()).getContainer());
+    const tooltipDiv = getTooltipDiv(container);
+    selection
+        .filter(d => d.baseValue !== d.mainValue)
+        .on("mouseover", function(data) {
+            generateHtml(tooltipDiv, data, settings);
+            showTooltip(container.node(), this, tooltipDiv);
+        })
+        .on("mouseout", function() {
+            hideTooltip(tooltipDiv);
+        });
+}
+
+function getTooltipDiv(container) {
+    let tooltipDiv = container.select("div.tooltip");
+    if (tooltipDiv.size() === 0) {
+        tooltipDiv = container
+            .append("div")
+            .attr("class", "tooltip")
+            .style("z-index", 3)
+            .style("opacity", 0)
+            .html(tooltipTemplate);
+    }
+    return tooltipDiv;
+}
+
+function generateHtml(tooltipDiv, data, settings) {
+    const splits = data.key.split("|");
+
+    // Add group data
+    if (settings.crossValues.length) {
+        const groups = data.crossValue.split(",");
+        tooltipDiv
+            .select("#cross-values")
+            .selectAll("li")
+            .data(groups)
+            .join("li")
+            .text((_, i) => `${settings.crossValues[i].name}: `)
+            .append("b")
+            .text(d => d);
+    }
+
+    // Add split data
+    if (settings.splitValues.length) {
+        tooltipDiv
+            .select("#split-values")
+            .selectAll("li")
+            .data(splits.slice(0, -1))
+            .join("li")
+            .text((_, i) => `${settings.splitValues[i].name}: `)
+            .append("b")
+            .text(d => d);
```

Comment:
Is there a way I could combine these `.text` calls? It's not a huge issue since there should never be that many groups or splits but it might be nicer to join them together.

Reply:
Perhaps?

````javascript
.join("li")
.html((d, i) =>`${settings.splitValues[i].name}: <b>d</b>`)
````

I'm not loving the html in a string, but it looks like it should work.

Reply:
I tried to avoid doing that by using a `.each`, I've commited it just now